consensus/src/pos/target.rs

```diff
@@ -0,0 +1,566 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chainstate_types::{BlockIndex, BlockIndexHandle};
+use common::{
+    chain::{
+        block::ConsensusData, ChainConfig, GenBlock, GenBlockId, PoSChainConfig, PoSStatus,
+        RequiredConsensus,
+    },
+    primitives::{BlockDistance, BlockHeight, Compact, Id},
+    Uint256, Uint512,
+};
+
+use crate::pos::error::ConsensusPoSError;
+
+fn calculate_average_block_time(
+    chain_config: &ChainConfig,
+    pos_config: &PoSChainConfig,
+    block_index: &BlockIndex,
+    block_index_handle: &impl BlockIndexHandle,
+) -> Result<u64, ConsensusPoSError> {
+    // Determine net upgrade version that current block belongs to.
+    // Across versions config parameters might change so it's invalid to mix versions for target calculations
+    let (_, net_version) = chain_config
+        .net_upgrade()
+        .version_at_height(block_index.block_height())
+        .expect("NetUpgrade must've been initialized");
+    let net_version_range = chain_config
+        .net_upgrade()
+        .height_range(net_version)
+        .expect("NetUpgrade must've been initialized");
+
+    // Average is calculated based on 2 timestamps and then is divided by number of blocks in between.
+    // Choose a block from the history that would be the start of a timespan.
+    // It shouldn't cross net version range or genesis.
+    let timespan_start_height = std::cmp::max(
+        net_version_range.start,
+        (block_index.block_height()
+            - BlockDistance::new(pos_config.block_count_to_average_for_blocktime() as i64))
+        .unwrap_or(BlockHeight::zero()),
+    );
+
+    let time_span_start = block_index_handle
+        .get_ancestor(block_index, timespan_start_height)?
+        .block_timestamp()
+        .as_int_seconds();
+    let current_block_time = block_index.block_timestamp().as_int_seconds();
+
+    let blocks_in_timespan: i64 = (block_index.block_height() - timespan_start_height)
+        .expect("cannot be negative")
+        .into();
+
+    let average = (current_block_time - time_span_start) / blocks_in_timespan as u64;
+
+    Ok(average)
+}
+
+fn calculate_new_target(
+    pos_config: &PoSChainConfig,
+    prev_target: &Uint256,
+    average_block_time: u64,
+) -> Result<Compact, ConsensusPoSError> {
+    let average_block_time = Uint512::from_u64(average_block_time);
+    let target_block_time = Uint512::from_u64(pos_config.target_block_time().get());
+    let prev_target: Uint512 = (*prev_target).into();
+
+    // TODO: limiting factor (mintlayer/mintlayer-core#787)
+    let new_target = prev_target * average_block_time / target_block_time;
+    let new_target = Uint256::try_from(new_target)?;
```

Comment:
There's a min difficulty to protect from this.

Reply:
Ah I see now. In that case it makes sense to me to saturate if `new_target` happens to be over `Uint256::MAX`, rather than return an error. Since it is going to be "saturated" at the min difficulty target level anyway.

Reply:
sounds reasonable, done